File: plugin/src/main/java/de/zoeyvid/ytparty/party/Party.java
package de.zoeyvid.ytparty.party;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public final class Party {
    public record TrackRef(int id, String uri, String title, String requester) {}
    public record MgrReport(long pos, long at) {}

    public final String id;
    public final Map<UUID, PermissionLevel> members = new LinkedHashMap<>();
    public final Map<UUID, PermissionLevel> invites = new HashMap<>();
    public final List<TrackRef> tracks = new ArrayList<>();
    public final Map<UUID, MgrReport> mgrPos = new HashMap<>();
    public boolean isPublic;
    public PermissionLevel publicJoinLevel;
    public int currentIndex = -1;
    public boolean paused = false;
    public boolean autoRemovePlayed = true;
    public byte sbFlags = 0x0F;
    public boolean repeatOne = false;
    public int nextTrackId = 1;
    public int generation = 0;

    public Party(String id, UUID host, boolean isPublic, PermissionLevel publicJoinLevel) {
        this.id = id;
        this.isPublic = isPublic;
        this.publicJoinLevel = publicJoinLevel;
        members.put(host, PermissionLevel.MANAGE);
    }

    public PermissionLevel level(UUID u) { return members.getOrDefault(u, PermissionLevel.LISTEN); }
    public boolean canManage(UUID u) { return level(u).canManage(); }
    public boolean canInvite(UUID u) { return level(u).canInvite(); }

    public int curTrackId() { return currentIndex >= 0 && currentIndex < tracks.size() ? tracks.get(currentIndex).id() : 0; }

    public int indexOf(int id) {
        for (int i = 0; i < tracks.size(); i++) if (tracks.get(i).id() == id) return i;
        return -1;
    }

    public boolean hasManager() {
        for (PermissionLevel l : members.values()) if (l.canManage()) return true;
        return false;
    }

    public void move(int from, int to) {
        if (from < 0 || from >= tracks.size() || to < 0 || to >= tracks.size()) return;
        tracks.add(to, tracks.remove(from));
    }
}

File: relay/party.go
package main

import (
	"slices"
	"strings"
)

const (
	listen = 0
	invite = 1
	manage = 2
)

func levelFromName(s string) int {
	switch strings.ToLower(s) {
	case "manage":
		return manage
	case "invite":
		return invite
	default:
		return listen
	}
}
func lvl(b int) int {
	if b < listen || b > manage {
		return listen
	}
	return b
}
func capped(level, mx int) int {
	if level > mx {
		return mx
	}
	return level
}

type party struct {
	id          string
	members     map[string]int
	invites     map[string]int
	tracks      []trackRef
	isPublic    bool
	pubJoinLvl  int
	curIndex    int
	paused      bool
	autoRemove  bool
	sbFlags     byte
	repeatOne   bool
	nextTrackId int
	generation  int
	trackStart  int64
	pausedAccum int64
	pausedSince int64
}

type trackRef struct {
	id    int
	uri   []byte
	title []byte
	reqr  []byte
}

func (p *party) curTrackID() int {
	if p.curIndex >= 0 && p.curIndex < len(p.tracks) {
		return p.tracks[p.curIndex].id
	}
	return 0
}

func (p *party) indexOf(id int) int {
	for i := range p.tracks {
		if p.tracks[i].id == id {
			return i
		}
	}
	return -1
}

func (p *party) level(tok string) int {
	if l, ok := p.members[tok]; ok {
		return l
	}
	return listen
}
func (p *party) hasManager() bool {
	for _, l := range p.members {
		if l == manage {
			return true
		}
	}
	return false
}
func (p *party) move(from, to int) {
	if from < 0 || from >= len(p.tracks) || to < 0 || to >= len(p.tracks) {
		return
	}
	t := p.tracks[from]
	p.tracks = slices.Delete(p.tracks, from, from+1)
	p.tracks = slices.Insert(p.tracks, to, t)
}

func nilGet(p *party, tok string) (int, bool) {
	if p == nil {
		return 0, false
	}
	v, ok := p.members[tok]
	return v, ok
}
func nilInv(p *party, name string) bool {
	if p == nil {
		return false
	}
	_, ok := p.invites[name]
	return ok
}
